test/TestUtils.ts

```diff
@@ -0,0 +1,426 @@
+
+type StartRelayParams = {
+  serverConfig: ServerConfigParams;
+  delay?: number;
+  stake?: BigNumberish;
+  relayOwner: string;
+  relayHubAddress: string;
+};
+
+const provider = ethers.provider;
+
+const startRelay = async (options: StartRelayParams) => {
+  const { serverConfig, delay, stake, relayOwner, relayHubAddress } = options;
+
+  fs.rmSync(SERVER_WORK_DIR, { recursive: true, force: true });
+
+  const originalConfig = { ...config };
+
+  config.util.extendDeep(originalConfig, {
+    ...serverConfig,
+    contracts: {
+      relayHubAddress,
+    },
+    app: {
+      workdir: SERVER_WORK_DIR,
+      devMode: true,
+      checkInterval: 10,
+      logLevel: 5,
+    },
+  });
+
+  const url = buildServerUrl();
+
+  const runServerPath = path.resolve(
+    __dirname,
+    '../node_modules/@rsksmart/rif-relay-server/dist/commands/Start.js'
+  );
+
+  const proc: ChildProcessWithoutNullStreams & { alreadyStarted?: number } =
+    childProcess.spawn('node', [runServerPath]);
+
+  const { relayManagerAddress } = await waitForFunction(
+    getServerStatus,
+    url,
+    "can't ping server",
+    3
+  );
+
+  console.log('Relay Server Manager Address', relayManagerAddress);
+
+  await provider.getSigner(relayOwner).sendTransaction({
+    to: relayManagerAddress,
+    value: utils.parseEther('2'),
+  });
+
+  const relayHub = RelayHub__factory.connect(relayHubAddress, provider);
+
+  await relayHub.stakeForAddress(relayManagerAddress, delay || 2000, {
+    from: relayOwner,
+    value: stake || utils.parseEther('1'),
+  });
+
+  const { relayWorkerAddress } = await waitForFunction(
+    getServerReady,
+    url,
+    'timed out waiting for relay to get staked and registered',
+    25,
+    500
+  );
+
+  return {
+    proc,
+    worker: relayWorkerAddress,
+    manager: relayManagerAddress,
+  };
+};
+
+const buildServerUrl = () => {
+  const app = config.get<AppConfig>('app');
+
+  const portFromUrl = app.url.match(/:(\d{0,5})$/);
+
+  return !portFromUrl && app.port ? `${app.url}:${app.port}` : app.url;
+};
+
+const getServerStatus = async (
+  url: string,
+  client = new HttpClient(new HttpWrapper(undefined, 'silent'))
+): Promise<HubInfo> => client.getChainInfo(url);
+
+const getServerReady = async (url: string): Promise<HubInfo | undefined> => {
+  const response = await getServerStatus(url);
+
+  if (response.ready) {
+    return response;
+  }
+
+  return undefined;
+};
+
+const waitForFunction = async (
+  functionToWait: (url: string) => Promise<HubInfo | undefined>,
+  url: string,
+  errorMessage: string,
+  attempts: number,
+  interval = 1000
+): Promise<HubInfo> => {
+  for (let i = 0; i < attempts; i++) {
+    try {
+      console.log('sleep before cont.');
+      await sleep(interval);
+      const response = await functionToWait(url);
+
+      if (!response) {
+        continue;
+      }
+
+      return response;
+    } catch (e) {
+      console.log('getChainInfo failed');
+    }
+  }
+
+  throw Error(errorMessage);
+};
```

Comment:
I'd rename and simplify to:

```suggestion
const waitForHubInfoProperty = async (
  url: string,
  hubInfoProp: keyof HubInfo,
  attempts: number,
  interval = 1000,
  customErrorMessage?: string,
): Promise<HubInfo> => {
  const client = new HttpClient(new HttpWrapper(undefined, 'silent'));
  for (let i = 0; i < attempts; i++) {
    try {
      console.info('sleep before cont.');
      await sleep(interval);

      const hubInfo: HubInfo | undefined = await client.getChainInfo(url);

      if (hubInfo?.[hubInfoProp]) {
        continue;
      }

      return hubInfo;
    } catch (error: unknown) {
      console.error(`Failed to call getChainInfo on ${url}:`, error);
    }
  }

  throw Error(`${customErrorMessage && customErrorMessage + ': '}HubInfo poll expired and ${hubInfoProp} property was not aquired from ${url}.getChainInfo`);
};
```

as it seems to better capture the nature of what it actually does.

Reply:
updated